tool/tctl/common/bots_command.go

```diff
@@ -483,7 +498,165 @@ func (c *BotsCommand) LockBot(ctx context.Context, client auth.ClientI) error {
 	return nil
 }
 
-func splitRoles(flag string) []string {
+// updateBotLogins applies updates from CLI arguments to a bot's logins trait,
+// updating the field mask if any updates were made.
+func (c *BotsCommand) updateBotLogins(ctx context.Context, bot *machineidv1pb.Bot, mask *fieldmaskpb.FieldMask) error {
```

Comment:
It would be great to add tests for both updateBotRoles and updateBotLogins. The logins func should be straightforward. The roles func could be easy if the code that created desired roles was isolated in its own func (so you don’t have to include auth.ClientI in the test). updateBotRoles could call the new func to get desired roles then call the client. 

The other thing to consider is that this code may be simplified by not using maps at all. Both funcs essentially follow a similar pattern. Compare the current unique list of things to a new unique list of things and set the new list if they are different. Since all the things (logins and roles) are already lists and we need to create a new list, sticking with a sorted slice has the potential to reduce complexity. We’re not dealing with a huge amount of items in each list so performance of a sorting algorithm should be acceptable. 

Here’s an untested and naive example for the logins func. LMK what you think.

```go
func (c *BotsCommand) updateBotLogins(ctx context.Context, bot *machineidv1pb.Bot, mask *fieldmaskpb.FieldMask) error {
	traits := bot.Spec.GetTraits()

	index := -1
	var current []string
	for i, t := range traits { // []*Trait
		if t.Name == constants.TraitLogins {
			index = i
			current = t.Values
			break
		}
	}
	sort.Strings(current)

	desired := uniqStrings(splitEntries(c.setLogins)) // sorted []string
	if len(desired) == 0 {
		desired = slices.Clone(current)
	}
	desired = uniqStrings(append(desired, splitEntries(c.addLogins)...))

	if slices.Equal(current, desired) {
		return nil
	}

	if len(desired) > 0 {
		traits[index].Values = desired
	} else if len(traits) > 0 && index != -1 {
		traits = append(traits[:index], traits[index+1:]...)
	}
	bot.Spec.Traits = traits

	return trace.Wrap(mask.Append(&machineidv1pb.Bot{}, "spec.traits"))
}

func uniqStrings(list []string) []string {
	sort.Strings(list)

	var uniq []string
	for i := range list {
		if i == 0 || list[i] != list[i-1] {
			uniq = append(uniq, list[i])
		}
	}
	return uniq
}
```

Reply:
> It would be great to add tests for both updateBotRoles and updateBotLogins.

Sure, I've added tests for both functions. It was pretty straightforward to test them in situ with a minimal client interface, so the structure is mostly left the same.

> The other thing to consider is that this code may be simplified by not using maps at all.

I do agree this approach is reasonable, but I think there is precedent for using maps here to deduplicate and compare the lists, and would probably consider it the idiomatic way of doing so in Go. We have [hundreds](https://github.com/search?q=repo%3Agravitational%2Fteleport+%2Fmap%5C%5Bstring%5C%5Dstruct%5C%7B%5C%7D%2F&type=code) of string sets throughout Teleport, many of which are used to deduplicate and compare string lists (and including several specialized `set` implementations).

That said, while searching I did discover the preexisting `utils.StringsSliceFromSet()` which reduces LoC a bit, and sorts the log output, which is nice for end users.